chainer/dataset/download.py

```diff
@@ -0,0 +1,148 @@
+
+    Returns:
+        str: Path to the dataset directory.
+
+    """
+    path = os.path.join(_dataset_root, dataset_name)
+    if create_directory:
+        try:
+            os.makedirs(path)
+        except OSError:
+            pass
+    return path
+
+
+def cached_download(url):
+    """Downloads a file and caches it.
+
+    Args:
+        url (str): URL to download from.
+
+    Returns:
+        str: Path to the downloaded file.
+
+    """
+    cache_root = os.path.join(_dataset_root, '_dl_cache')
+    try:
+        os.makedirs(cache_root)
+    except OSError:
+        if not os.path.exists(cache_root):
+            raise RuntimeError('cannot create download cache directory')
+
+    lock_path = os.path.join(cache_root, '_dl_lock')
+    urlhash = hashlib.md5(url.encode('utf-8')).hexdigest()
+    cache_path = os.path.join(cache_root, urlhash)
+
+    with filelock.FileLock(lock_path):
+        if os.path.exists(cache_path):
+            return cache_path
+
+    temp_root = tempfile.mkdtemp(dir=cache_root)
+    try:
+        temp_path = os.path.join(temp_root, 'dl')
+        print('Downloading from {}...'.format(url))
+        request.urlretrieve(url, temp_path)
+        with filelock.FileLock(lock_path):
+            shutil.move(temp_path, cache_path)
+    finally:
+        shutil.rmtree(temp_root)
+
+    return cache_path
+
+
+def cache_or_load_file(path, creator, loader):
+    """Caches a file if it does not exist, or loads it otherwise.
+
+    This is a utility function used in dataset loading routines. The creator
+    creates the file to given path, and returns the content. If the file
+    already exists, the loader is called instead, and it loads the cached file
+    and returns the content.
+
+    Note that the path passed to the creator is temporary one, and not same as
+    the path given to this function. This function safely renames the file
+    created by the creator to a given path, even if this function is called
+    simultaneously by multiple threads or processes.
+
+    Args:
+        path (str): Path to save the cached file.
+        creator: Function to create the file and returns the content. It takes
+            a path to temporary place as the argument. Before calling the
+            creator, there is no file at the temporary path.
+        loader: Function to load the cached file and returns the content.
+
+    Returns:
+        It returns the returned values by the creator or the loader.
+
+    """
+    if os.path.exists(path):
+        return loader(path)
+
+    file_name = os.path.basename(path)
+    temp_dir = tempfile.mkdtemp()
+    temp_path = os.path.join(temp_dir, file_name)
+
+    lock_path = os.path.join(_dataset_root, '_create_lock')
+
+    try:
+        content = creator(temp_path)
+        with filelock.FileLock(lock_path):
+            if not os.path.exists(path):
+                shutil.move(temp_path, path)
+    except OSError:
+        if not os.path.exists(path):
+            raise
```

Comment:
I forgot the implication behind this code... maybe I did not assume that `creator` raises `OSError`, though it would actually raise it. I will fix this code.

> What happens creator raises an error other than OSError?

In this case, the code just pass through the error.